HDR images don't get dimmed when Safari goes to background

mwyrzykowski opened the pull request:
#### 460ef7c5d3339ab16a256bfaa2b9dbc3e9f9d34e
<pre>
HDR images don&apos;t get dimmed when Safari goes to background
<a href="https://bugs.webkit.org/show_bug.cgi?id=294092">https://bugs.webkit.org/show_bug.cgi?id=294092</a>
<a href="https://rdar.apple.com/151892044">rdar://151892044</a>

Reviewed by Gerald Squelart.

We were failing to adjust the preferred dynamic range of the layer
during layer updates, resulting in mistmatched appearances during
background / foregrounding.

* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::updateContentsFormat):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::setDynamicRangeLimitRecursive):
(WebKit::setDynamicRangeLimit):
(WebKit::RemoteLayerTreeHost::layerWillBeRemoved):
Largely moved from WebViewImpl

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setShouldSuppressHDR):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::applicationShouldSuppressHDR):
(WebKit::setDynamicRangeLimitRecursive): Deleted.
(WebKit::setDynamicRangeLimit): Deleted.
Moved to RemoteLayerTreeHost.

* Source/WebCore/platform/graphics/PlatformDynamicRangeLimit.h:
(WebCore::PlatformDynamicRangeLimit::defaultWhenSuppressingHDR):
Default to constrained now.

* Tools/TestWebKitAPI/Tests/WebCore/PlatformDynamicRangeLimitTests.cpp:
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, Values)):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, StaticValues)):
Update expected values.

Canonical link: <a href="https://commits.webkit.org/296171@main">https://commits.webkit.org/296171@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8517caee0ae49f5c7cd8b6afe0dbfd817967501

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107546 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27229 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112763 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58088 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109510 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27920 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35730 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81649 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110475 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22126 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96935 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62030 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21564 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15074 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57526 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91496 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15108 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115864 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34614 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25526 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90690 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34990 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93186 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90431 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23068 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35344 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13125 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30377 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34535 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40091 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34281 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37640 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35944 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -1250,11 +1250,9 @@ static bool isInRecoveryOS()
-    if ([layer wantsExtendedDynamicRangeContent]) {
-    ALLOW_DEPRECATED_DECLARATIONS_END
+    if ([layer toneMapMode] == CAToneMapModeIfSupported)
         layerDynamicRangeLimitSetter(layer);
```
this is where we set CADynamicRangeStandard in the UI process @smfr

squelart replied:
So we need to keep `WebViewImpl::applicationShouldSuppressHDR()` (above in this file) and `GraphicsLayerCA::hdrContentsFormat()` absolutely in sync, so that they always do the same thing when suppressing HDR or not!
Would it be possible to move this logic in one common place? (Maybe in a follow-up bug, in which case some scary comments would be nice in this patch, so we don't modify one and not the other by mistake.)

mwyrzykowski replied:
It would be great to unify that, I was not sure how given my unfamiliarity. Canvas layers don't seem to take the `GraphicsLayerCA::hdrContentsFormat()` path for instance

squelart replied:
Yeah, it's probably not trivial. So I'm personally fine with landing this (with scary comments!) first, to fix the problem, and you/I/someone can revisit the logic later.

mwyrzykowski replied:
I ended up changing the approach so no need for the comment I think?

squelart commented on Source/WebCore/platform/graphics/ca/cocoa/ContentsFormatCocoa.h:
```diff
@@ -64,11 +66,39 @@ constexpr NSString *contentsFormatString(ContentsFormat contentsFormat)
+        [layer setPreferredDynamicRange:CADynamicRangeConstrainedHigh];
+        break;
+    case ContentsFormat::RGBA16FSDR:
+        [layer setPreferredDynamicRange:CADynamicRangeStandard];
```
You should probably guard these layer calls with `#if HAVE(SUPPORT_HDR_DISPLAY_APIS)`.

squelart reviewed (approved):
r+ with suggested guard.
(And assuming it works correctly; it may impact non-HDR pixels in layers, that may require fixing in a separate PR.)

shallawa commented on Source/WebCore/platform/graphics/ContentsFormat.h:
```diff
@@ -42,6 +42,8 @@ enum class ContentsFormat : uint8_t {
 #if ENABLE(PIXEL_FORMAT_RGBA16F)
     RGBA16F = 1 << 2,
+    RGBA16FConstrained = 1 << 3,
+    RGBA16FSDR = 1 << 4,
```
What is the meaning of `RGBA16F` after adding `RGBA16FConstrained` and `RGBA16FSDR`? Does it actually mean `RGBA16FHigh` or `RGBA16FNoLimit`? If yes, then these new enum are like combining `ContentsFormat` and `DynamicRangeLimit` (without the mix function) in one enum type.

I think `ContentsFormat` should describe the backing store of the layer only. It should describe whether is 8, 10 or 16 bits per channel. It should not include other information about tone-mapping the pixels to a certain `DynamicRangeLimit`.

Can't we add another member of type `DynamicRangeLimit` or `PlatformDynamicRangeLimit` to control suppressing the HDR-ness of the `RGBA16F` layer?

mwyrzykowski replied:
Yes that is certainly a good option. I wasn’t sure if we needed to keep the sizeof(LayerProperties) small or if the additional enum byte is fine.

I’ll refactor the patch to add DynamicRangeLimit

smfr reviewed (changes requested):
I don't think we should put this into `ContentsFormat`. Keep it in the UI process as we discussed.

jesxilin commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
```diff
@@ -162,6 +163,35 @@
+        setLayerDynamicRangeLimit(layer, platformDynamicRangeLimit);
+
+    for (CALayer* sublayer in [layer sublayers])
+        setLayerDynamicRangeLimit(sublayer, platformDynamicRangeLimit);
```
This function doesn't actually recurse. Is this line supposed to be `setDynamicRangeLimitRecursive(sublayer, platformDynamicRangeLimit);` or perhaps a different function name?

mwyrzykowski replied:
oh nice observation 👍 , it does eventually recurse because `setDynamicRangeLimit -> setDynamicRangeLimitRecursive -> (for each sublayer) setDynamicRangeLimit -> setDynamicRangeLimitRecursive ...` but I can see how that is confusing (and was unintentional).

`setDynamicRangeLimit` can just be removed I think

pxlcoder replied:
Note that `setLayerDynamicRangeLimit` is called here, not `setDynamicRangeLimit` – so I think Jessica is right about a lack of recursion (and a likely logic error)?

mwyrzykowski replied:
oh indeed that is correct, I completely missed that, thank you

jesxilin commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
```diff
@@ -162,6 +163,35 @@
 #endif
 
+#if HAVE(SUPPORT_HDR_DISPLAY_APIS)
+static void setDynamicRangeLimitRecursive(CALayer* layer, PlatformDynamicRangeLimit platformDynamicRangeLimit)
```
`CALayer` is an Objective-C type, so I believe all `CALayer* layer` instances should be `CALayer *layer` instead.

mwyrzykowski replied:
oh had no idea about this - thank you! I see its quite inconsistent throughout the codebase (`CALayer *` vs `CALayer*`) but I will switch it

mwyrzykowski replied:
do you have the reference for this @jesxilin? I don't see it listed https://webkit.org/code-style-guidelines/

cdumez replied:
I don't know if it's written anywhere but Jessica is right about the style for ObjC types.

mwyrzykowski replied:
Ah I think I found it :) 
https://webkit.org/code-style-guidelines/#pointers-non-cpp

I presume 'non-C++' code in this wording is referring to Objective-C/C++

cdumez commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
```diff
@@ -162,6 +163,35 @@
+
+    PlatformDynamicRangeLimit platformDynamicRangeLimit = page->shouldSuppressHDR() ? PlatformDynamicRangeLimit::defaultWhenSuppressingHDR() : PlatformDynamicRangeLimit::noLimit();
+    [CATransaction begin];
+    for (auto [_, node] : m_nodes)
```
We normally iterate over `m_nodes.values()` instead of this `[_, node]` trick.

mwyrzykowski replied:
I only need to call this on `rootLayer()`, so I made that change instead

mwyrzykowski committed:
HDR images don't get dimmed when Safari goes to background
https://bugs.webkit.org/show_bug.cgi?id=294092
rdar://151892044

Reviewed by Gerald Squelart.

We were failing to adjust the preferred dynamic range of the layer
during layer updates, resulting in mistmatched appearances during
background / foregrounding.

* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::updateContentsFormat):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::setDynamicRangeLimitRecursive):
(WebKit::setDynamicRangeLimit):
(WebKit::RemoteLayerTreeHost::layerWillBeRemoved):
Largely moved from WebViewImpl

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setShouldSuppressHDR):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::applicationShouldSuppressHDR):
(WebKit::setDynamicRangeLimitRecursive): Deleted.
(WebKit::setDynamicRangeLimit): Deleted.
Moved to RemoteLayerTreeHost.

* Source/WebCore/platform/graphics/PlatformDynamicRangeLimit.h:
(WebCore::PlatformDynamicRangeLimit::defaultWhenSuppressingHDR):
Default to constrained now.

* Tools/TestWebKitAPI/Tests/WebCore/PlatformDynamicRangeLimitTests.cpp:
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, Values)):
(TestWebKitAPI::TEST(PlatformDynamicRangeLimit, StaticValues)):
Update expected values.

Canonical link: https://commits.webkit.org/296171@main
